hive/agents/ppo.py

```diff
@@ -0,0 +1,411 @@
+        if (self._replay_buffer.ready()):
+            self._replay_buffer.compute_advantages(self._value)
+            #TODO: recheck sampling
+            for _ in range(self._num_epochs):
+                valid_ind_size = self._replay_buffer._find_valid_indices()
+                for _ in range(valid_ind_size//self._batch_size):
+                    batch = self._replay_buffer.sample(batch_size=self._batch_size)
+                    batch = self.preprocess_update_batch(batch)
+                    self._actor_optimizer.zero_grad()
+                    self._critic_optimizer.zero_grad()
+
+                    _, _logprob, entropy = self._actor(batch['observation'], batch["action"])
+                    values = self._critic(batch['observation'])
+                    logratios = _logprob - batch["logprob"]
+                    ratios = torch.exp(logratios) 
+                    advantages = batch["advantages"]
+                    if self._normalize_advantages:
+                        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+                    # Actor loss
+                    loss1 = -advantages * ratios
+                    loss2 = -advantages * torch.clamp(ratios, 1 - self._clip_coef, 1 + self._clip_coef)
+                    loss = torch.max(loss1, loss2).mean()
+                    entr_loss = entropy.mean()
+                    actor_loss = loss - self._ent_coef * entr_loss
+                    actor_loss.backward()
+                    if self._grad_clip is not None:
+                        torch.nn.utils.clip_grad_norm_(
+                            self._actor.parameters(), self._grad_clip
+                        )
+                    self._actor_optimizer.step()
+
+                    # Critic loss
+                    values = values.view(-1)
+                    if self._clip_vloss:
+                        v_loss_unclipped = self._critic_loss_fn(values, batch["returns"])
+                        v_clipped = batch["values"] + torch.clamp(values - batch["values"], 
+                            -self._clip_coef,
+                            self._clip_coef,
+                        )
+                        v_loss_clipped = self._critic_loss_fn(v_clipped, batch["returns"])
+                        v_loss_max = torch.max(v_loss_unclipped, v_loss_clipped)
+                        critic_loss = 0.5 * v_loss_max.mean()
+                    else:
+                        critic_loss = 0.5 * self._critic_loss_fn(values, batch["returns"]).mean()
+                    critic_loss = self._vf_coef * critic_loss
+                    critic_loss.backward()
+                    if self._grad_clip is not None:
+                        torch.nn.utils.clip_grad_norm_(
+                            self._critic.parameters(), self._grad_clip
+                        )
+                    self._critic_optimizer.step()
```

Comment:
When you optimize the actor, the gradient is being backpropogated through the representation network making the current value loss inconsistent. This will result in incorrect loss.